doc/sql.extensions/README.ddl.txt

```diff
@@ -603,3 +603,8 @@ ALTER TABLE <name> ... [ {ENABLE | DISABLE} PUBLICATION ]
 
 Defines whether replication is enabled for the specified table.
 If not specified in the CREATE TABLE statement, the database-level default behaviour is applied.
+
+24) COLLATE clause can be used as a part of character data type as per SQL standard.
+(Dmitry Sibiryakov)
+
+If is used twice, an error is returned.
```

Comment:
This does not correspond to changes below. For parameters you seem to allow both and use the latest one. What about local variables?

Reply:
This PR is intended for table fields only, I'm not sure if parameters and variables are covered. Should they?

Reply:
You have added the code to `ParameterClause::ctor` with a comment // Explicit collation has precedence over data type collation

And parameters/variables are declared using:
`column_domain_or_non_array_type collate_clause ...`
where `column_domain_or_non_array_type` now includes a `collate_clause` itself after your changes...
`domain_or_non_array_type -> domain_or_non_array_type_name -> non_array_type -> simple_type`

Reply:
This comment applies to cases when data type is derived from domain or column including collation. Old code silently reset collation to default if no explicit COLLATE was provided so I've added the condition and the comment.

Reply:
It is pity that setClause() is not available outside of parser. Can they be made static?

Reply:
> It is pity that setClause() is not available outside of parser. Can they be made static?

IMHO, they (along with `setClauseFlag` / `checkDuplicateClause` / `isDuplicateClause`) should be `static`, as they don't access the parser state. @asfernandes , do you agree?

Reply:
I would question the need in `collate_clause` after `domain_or_non_array_type` and `column_domain_or_non_array_type`, as now it's also included inside these parser rules. The only regression would be that `NOT NULL` must be written after `COLLATE`, not before as now. Is it an acceptable breakage, given the low probability that both co-exist in input/output parameters / return values in real-world scripts? And there will be no need to check for duplicated definitions then.

Reply:
If nodes are going to call `Parser::setClause`, I don't agree. It's layer violation.
In this case I would move them to `Node` class.

Reply:
Collation can be removed from ParameterClause constructor parameters and assigned later. Is it ok?